api/test/trace/propagation/b3_propagation_test.cc

```diff
@@ -0,0 +1,138 @@
+}
+
+TEST(B3PropagationTest, TraceFlagsBufferGeneration)
+{
+  EXPECT_EQ(MapB3Context::GenerateTraceFlagsFromString("0"), trace::TraceFlags());
+}
+
+TEST(B3PropagationTest, PropagateInvalidContext)
+{
+  // Do not propagate invalid trace context.
+  std::map<std::string, std::string> carrier = {};
+  context::Context ctx{
+      "current-span",
+      nostd::shared_ptr<trace::Span>(new trace::DefaultSpan(trace::SpanContext::GetInvalid()))};
+  format.Inject(Setter, carrier, ctx);
+  EXPECT_TRUE(carrier.count("b3") == 0);
+}
+
+TEST(B3PropagationTest, SetRemoteSpan)
+{
+  const std::map<std::string, std::string> carrier = {
+      {"b3", "80f198ee56343ba864fe8b2a57d3eff7-e457b5a2e4d86bd1-1-05e3ac9a4f6e3b90"}};
+  context::Context ctx1 = context::Context{};
+  context::Context ctx2 = format.Extract(Getter, carrier, ctx1);
+
+  auto ctx2_span = ctx2.GetValue(trace::kSpanKey);
+  EXPECT_TRUE(nostd::holds_alternative<nostd::shared_ptr<trace::Span>>(ctx2_span));
+
+  auto span = nostd::get<nostd::shared_ptr<trace::Span>>(ctx2_span);
+
+  EXPECT_EQ(Hex(span->GetContext().trace_id()), "80f198ee56343ba864fe8b2a57d3eff7");
+  EXPECT_EQ(Hex(span->GetContext().span_id()), "e457b5a2e4d86bd1");
+  EXPECT_EQ(span->GetContext().IsSampled(), true);
+  // EXPECT_EQ(span->GetContext().HasRemoteParent(), true); // TODO: unsure about this one
+}
+
+TEST(B3PropagationTest, GetCurrentSpan)
+{
+  constexpr uint8_t buf_span[]  = {1, 2, 3, 4, 5, 6, 7, 8};
+  constexpr uint8_t buf_trace[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16};
+  trace::SpanContext span_context{trace::TraceId{buf_trace}, trace::SpanId{buf_span},
+                                  trace::TraceFlags{true}, false};
+  nostd::shared_ptr<trace::Span> sp{new trace::DefaultSpan{span_context}};
+
+  // Set `sp` as the currently active span, which must be used by `Inject`.
+  trace::Scope scoped_span{sp};
+
+  std::map<std::string, std::string> headers = {};
+  format.Inject(Setter, headers, context::RuntimeContext::GetCurrent());
+  EXPECT_EQ(headers["b3"], "0102030405060708090a0b0c0d0e0f10-0102030405060708-1");
+}
+
+TEST(B3PropagationTest, GetCurrentSpanMultiHeader)
+{
+  constexpr uint8_t buf_span[]  = {1, 2, 3, 4, 5, 6, 7, 8};
+  constexpr uint8_t buf_trace[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16};
+  trace::SpanContext span_context{trace::TraceId{buf_trace}, trace::SpanId{buf_span},
+                                  trace::TraceFlags{true}, false};
+  nostd::shared_ptr<trace::Span> sp{new trace::DefaultSpan{span_context}};
+
+  // Set `sp` as the currently active span, which must be used by `Inject`.
+  trace::Scope scoped_span{sp};
+
+  std::map<std::string, std::string> headers = {};
+  formatMultiHeader.Inject(Setter, headers, context::RuntimeContext::GetCurrent());
+  EXPECT_EQ(headers["X-B3-TraceId"], "0102030405060708090a0b0c0d0e0f10");
+  EXPECT_EQ(headers["X-B3-SpanId"], "0102030405060708");
+  EXPECT_EQ(headers["X-B3-Sampled"], "1");
+}
```

Comment:
Would be nice to have a test of left-padding an incoming 64 bit trace id with zeroes